Add CSV batch inference tutorial (#136)
valohai/docs: source/tutorials/batch-inference/image-batch-inference.rst

.. code-block:: yaml

    ---

    - step:
        name: Batch Inference
        image: tensorflow/tensorflow:2.4.1
        command:
        - pip install pillow valohai-utils
        - python batch_inference.py
        inputs:
        - name: model
          default: s3://valohai-public-files/tutorials/batch-inference/image-batch-inference/model.h5
        - name: images
          default: s3://valohai-public-files/tutorials/batch-inference/image-batch-inference/images.zip

What we are doing here is defining a single step for our machine learning pipeline, which is the Batch Inference step. We will run on top of the official ``tensorflow/tensorflow:2.4.1`` Docker image, first install the ``pillow`` and ``valohai-utils`` Python libraries and then run our batch inference code.

Let's test that everything is set up correctly by running on Valohai:

.. code-block:: bash

    vh exec run --adhoc "Batch Inference"

If everything went as planned, we should see our Valohai execution end after finding out that ``batch_inference.py`` is missing:

.. image:: batch-inference-tutorial-1.png
   :alt: Error, but success!

Unpacking the Images
--------------------

> ...actually Valohai unpacks the images automatically, so we can just fetch their location with ``vh.inputs('images').paths()``!

Reviewer: This should be explained better. It's not Valohai unpacking images, it's the `valohai-utils` library, when `process_archives` is on...

Loading and Using Our Model
---------------------------

Okey, let's get started by creating and opening up ``batch_inference.py`` in your favorite editor!

Add these imports to the beginning of the file:

.. code-block:: python

    import json
    import os

    import numpy as np
    from PIL import Image
    import tensorflow as tf
    import valohai as vh

We are using the ``valohai-utils`` library to make integration with Valohai straightforward.

Begin by loading our model:

.. code-block:: python

    model = tf.keras.models.load_model(vh.inputs('model').path())

> Easy, huh? Let's define a function to load up an image:

Reviewer: ```suggestion
Easy, huh? Let's define a function to load up an image.

We convert the loaded image to grayscale (Pillow mode `L`), then resize it to 28x28 pixels,
since that's the size our model expects. Finally, the image levels are normalized from 0..255 to 0..1.
```

.. code-block:: python

    def load_image(image_path):

>         _, image_name = os.path.split(image_path)

Reviewer: ```suggestion
        image_name = os.path.basename(image_path)
```

        image = Image.open(image_path)

>         image.load()

Reviewer: I'm not sure this is necessary.

        image = image.resize((28, 28)).convert('L')
        image_data = np.array(image).reshape(1, 28, 28)
        image_data = image_data / 255.0
        return (image_name, image_data)

Then define a function to run inference on an image:

.. code-block:: python

    def run_inference(image):
        image_name, image_data = image
        prediction = np.argmax(model.predict(image_data))
        with vh.logger() as logger:

>             logger.log('image', image_name)
>             logger.log('inferred_digit', prediction)

Reviewer: ```suggestion
            logger.log(image=image_name, inferred_digit=prediction)
```

Author: This is a good suggestion, but I think I'll stick with the simple version.

Reviewer: I think this is simpler :shrug:

        return {
            'image': image_name,
            'inferred_digit': str(prediction),
        }

Finally, let's run these functions for all the images. While we already log all the individual inference results in the inference function, let's also save the complete results in a single JSON file.

.. code-block:: python

>     image_paths = vh.inputs('images').paths()
> 
>     images = [load_image(p) for p in image_paths]
>     results = [run_inference(i) for i in images]

Reviewer: This might not be a good idea for a large dataset, since you're loading all images into memory first, then processing them.

```suggestion
    results = []
    for path in vh.inputs('images').paths():
        results.append(run_inference(load_image(path)))
```

Author: We clearly need lazy evaluation! 😄

Reviewer: Well, sure:

```
results = [run_inference(i) for i in (load_image(p) for p in vh.inputs('images').paths())]
```

That makes the load-image bit a genexpr, which is lazy. It's not readable, though...

Author: 😄 yeah, I think your original suggestion is better.

    with open(vh.outputs().path('results.json'), 'w') as f:
        json.dump(results, f)

Let's run the batch inference on Valohai:

.. code-block:: bash

    vh exec run --adhoc "Batch Inference"

> If everything went according to plan, you can now preview the results in the Outputs-tab:

Reviewer: ```suggestion
If everything went according to plan, you can now preview the results in the Outputs tab:
```
